File: 0538/0538.cpp
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};


class Solution {
    int sum = 0;
public:
    TreeNode* convertBST(TreeNode* root) {
        dfs(root);
        return root;
    }

    int dfs(TreeNode* node) {
        if (node == nullptr) {
            return 0;
        }
        dfs(node->right);
        sum += node->val;
        node->val = sum;
        dfs(node->left);
        return 0;
    }
};

File: 0538/0538.py
'''
Author       : Zhu Honglin
Date         : 2020-09-21 12:58:47
LastEditTime : 2020-09-21 14:10:11
'''

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def convertBST(self, root: TreeNode) -> TreeNode:
        self.total_sum = 0
        
        def dfs(node):
            if not node:
                return
            dfs(node.right)
            self.total_sum += node.val
            node.val = self.total_sum
            dfs(node.left)

        dfs(root)
        return root
    
        